src/k2/module/k23si/Module.cpp

```diff
@@ -256,34 +265,41 @@ dto::Key K23SIPartitionModule::_getContinuationToken(const IndexerIterator& it,
     }
     else if (it != _indexer.end()) {
         // This is the paginated case
+        request.exclusiveKey = false;
         return it->first;
     }
 
     // This is the multi-partition case
-    // TODO support this for reverseDirection scan
-    return dto::Key {
-        request.key.schemaName,
-        _partition().endKey,
-        ""
-    };
+    if (request.reverseDirection) {
+        request.exclusiveKey = true;
+        return dto::Key {
+            request.key.schemaName,
+            _partition().startKey,
+            ""
+        };
+    } else {
+        request.exclusiveKey = false;
+        return dto::Key {
+            request.key.schemaName,
+            _partition().endKey,
+            ""
+        };
+    }
 }
 
 seastar::future<std::tuple<Status, dto::K23SIQueryResponse>>
 K23SIPartitionModule::handleQuery(dto::K23SIQueryRequest&& request, dto::K23SIQueryResponse&& response, FastDeadline deadline) {
     K2DEBUG("Partition: " << _partition << ", received query " << request);
 
-    Status validateStatus = _validateReadRequest(request);
+    Status validateStatus = _validateQueryRequest(request);
     if (!validateStatus.is2xxOK()) {
         return RPCResponse(std::move(validateStatus), dto::K23SIQueryResponse{});
     }
     if (_partition.getHashScheme() != dto::HashScheme::Range) {
             return RPCResponse(dto::K23SIStatus::OperationNotAllowed("Query not implemented for hash partitioned collection"), dto::K23SIQueryResponse{});
     }
-    if (request.reverseDirection) {
-            return RPCResponse(dto::K23SIStatus::OperationNotAllowed("Reverse scan query not fully implemented"), dto::K23SIQueryResponse{});
-    }
 
-    IndexerIterator key_it = _initializeScan(request.key, request.reverseDirection);
+    IndexerIterator key_it = _initializeScan(request.key, request.reverseDirection, request.exclusiveKey);
 
     for (; !_isScanDone(key_it, request, response.results.size());
```

Comment:
If exclusiveKey is true in the request then the server should not return it in the result set